File: polars-rust/src/packs/unbalanced.rs
use polars::prelude::{col, JoinArgs, JoinType, LazyFrame};

/// Compute the Unbalanced JE Report
pub fn unbalanced_journal_entries_test(gl: &LazyFrame) -> LazyFrame {
    // Filter out 0 amount lines as they don't mean anything.
    let gl_filtered = gl.clone().filter(col("GL_Local_Amount").abs().gt(0.00001));

    // Get JE totals
    let je_totals = gl_filtered
        .clone()
        .group_by([col("GL_Journal_ID")])
        //TODO Adding the round funtion breaks data streaming, for now.
        //.agg([col("GL_Local_Amount").sum().round(2)]);
        .agg([col("GL_Local_Amount").sum()]);

    // Filter for unbalanced JEs
    let unbalanced_je = je_totals
        //TODO Adding the round funtion breaks data streaming, for now.
        //.filter(col("GL_Local_Amount").abs().gt(0));
        .filter(col("GL_Local_Amount").abs().gt(0.00001));

    // Pull all lines for unbalanced JEs
    let unbalanced_je_lines = gl_filtered.clone().join(
        unbalanced_je,
        [col("GL_Journal_ID")],
        [col("GL_Journal_ID")],
        JoinArgs::new(JoinType::Inner),
    );

    // Filter for just the columns we want on the report and sort them to make compression happier
    unbalanced_je_lines
        .select([
            col("GL_Journal_ID"),
            col("GL_Line_Number"),
            col("GL_Effective_Date"),
            col("GL_Account_Number"),
            col("GL_Local_Amount"),
        ])
        .sort_by_exprs(
            [col("GL_Journal_ID"), col("GL_Line_Number")],
            [false; 2],
            false,
            false,
        )
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::io::read_gl;
    use std::path::Path;

    #[test]
    fn does_unbalanced_run() {
        let gl = read_gl(Path::new("../gitignore/data/gl.parquet/gl_0000.parquet")).unwrap();

        let unbalanced: LazyFrame = unbalanced_journal_entries_test(&gl);
        let _ = unbalanced.collect().unwrap();
    }
}

File: pyspark/src/packs/unbalanced.py
from pyspark.sql.dataframe import DataFrame  # type: ignore
from pyspark.sql.functions import abs, round, sum  # type: ignore


# Compute the Unbalanced JE Report
def unbalanced_test(gl: DataFrame) -> DataFrame:
    # Push PySpark Dataframes to SQL like objects
    gl_filtered = gl.filter(gl["GL_Local_Amount"] != 0)

    # Get JE totals
    gl_je_total = gl_filtered.groupby("GL_Journal_ID").agg(
        sum("GL_Local_Amount").alias("GL_Local_Amount")
    )

    # Get Magnitude of all unbalanced JEs
    gl_je_magnitude = (
        gl_je_total.withColumn("GL_Local_Amount", abs("GL_Local_Amount"))
        .withColumn("GL_Local_Amount", round("GL_Local_Amount", 2))
        .withColumnRenamed("GL_Local_Amount", "GL_Magnitude")
    )

    # Get Unbalanced JEs
    gl_unbalanced_je = gl_je_magnitude.filter(gl_je_magnitude["GL_Magnitude"] > 0)

    # Get Unbalanced Lines
    gl_unbalanced_lines = gl_filtered.join(
        gl_unbalanced_je,
        on=["GL_Journal_ID"],
        how="inner",
    )

    # Filter just for the columns we want and sort so results compress well
    unbalanced_lines_report = gl_unbalanced_lines[
        [
            "GL_Journal_ID",
            "GL_Line_Number",
            "GL_Effective_Date",
            "GL_Account_Number",
            "GL_Local_Amount",
        ]
    ].sort(
        [
            "GL_Journal_ID",
            "GL_Line_Number",
        ]
    )
    return unbalanced_lines_report
